test/tests/functional/pbs_cray_suspend_resume.py

```diff
@@ -93,3 +97,390 @@ def test_exclusive_job_not_suspended(self):
+
+        self.server.expect(JOB, {ATTR_state: 'R', ATTR_substate: 42}, id=jid1)
+        self.server.expect(JOB, {ATTR_state: 'R', ATTR_substate: 42}, id=jid2)
+
+        qstat = self.server.status(JOB, 'exec_vnode', id=jid1)
+        vname1 = qstat[0]['exec_vnode'].partition(':')[0].strip('(')
+        qstat = self.server.status(JOB, 'exec_vnode', id=jid2)
+        vname2 = qstat[0]['exec_vnode'].partition(':')[0].strip('(')
+
+        # admin-suspend subjob 1
+        self.server.sigjob(jid1, 'admin-suspend', runas=ROOT_USER)
+        self.server.expect(JOB, {ATTR_state: 'S'}, id=jid1)
+        self.server.expect(NODE, {'state': 'maintenance'}, id=vname1)
+        self.server.expect(NODE, {'maintenance_jobs': jid1})
+
+        # admin-resume subjob 1 . Make sure the node retuns to state free
+        self.server.sigjob(jid1, 'admin-resume', runas=ROOT_USER)
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid1)
+        self.server.expect(NODE, {'state': 'job-exclusive'}, id=vname1)
+
+        # admin-suspend subjob 2 and restart
+        self.server.sigjob(jid2, 'admin-suspend', runas=ROOT_USER)
+        self.server.expect(JOB, {ATTR_state: 'S'}, id=jid2)
+        self.server.expect(NODE, {'state': 'maintenance'}, id=vname2)
+        self.server.expect(NODE, {'maintenance_jobs': jid2})
+        self.server.restart()
+        self.server.expect(NODE, {'state': 'free'}, id=vname2)
+        self.server.cleanup_jobs()
+
+    def test_admin_suspend_wrong_state(self):
+        """
+        Check that wrong 'resume' signal is correctly rejected.
+        """
+        j1 = Job(TEST_USER)
+        jid1 = self.server.submit(j1)
+        self.server.expect(JOB, {ATTR_state: 'R', ATTR_substate: 42}, id=jid1)
+        self.server.sigjob(jid1, "suspend", runas=ROOT_USER)
+        self.server.expect(JOB, {ATTR_state: 'S'}, id=jid1)
+
+        try:
+            self.server.sigjob(jid1, "admin-resume", runas=ROOT_USER)
+        except PbsSignalError as e:
+            self.assertTrue(
+                'Job can not be resumed with the requested resume signal'
+                in e.msg[0])
+        self.server.expect(JOB, {ATTR_state: 'S'}, id=jid1)
+
+        j2 = Job(TEST_USER)
+        jid2 = self.server.submit(j2)
+        self.server.expect(JOB, {ATTR_state: 'R', ATTR_substate: 42}, id=jid2)
+        self.server.sigjob(jid2, "admin-suspend", runas=ROOT_USER)
+        self.server.expect(JOB, {ATTR_state: 'S', ATTR_substate: 43}, id=jid2)
+
+        try:
+            self.server.sigjob(jid2, "resume", runas=ROOT_USER)
+        except PbsSignalError as e:
+            self.assertTrue(
+                'Job can not be resumed with the requested resume signal'
+                in e.msg[0])
+
+        # If resume had worked, the job would be in substate 45
+        self.server.expect(JOB, {ATTR_state: 'S', ATTR_substate: 43}, id=jid2)
+        self.server.cleanup_jobs()
+
+    def submit_resv(self, resv_start, chunks, resv_dur):
```

Comment:
I usually see helper functions at the top of the file before the tests start.  It seems like good convention, can you please move this up?  Also, it needs a doc string.

Reply:
I'll add the doc string.  I prefer to have the function right above the section that uses it. It seems easier to read without scrolling to the top of the file.

Reply:
Since there is no rule about where helper functions should go, this seems reasonable.